pkg/config/analysis/analyzers/envoyfilter/envoyfilter.go

```diff
@@ -45,33 +45,107 @@ func (*EnvoyPatchAnalyzer) Metadata() analysis.Metadata {
+			// or HTTP_ROUTE
+			if patch.ApplyTo == network.EnvoyFilter_ROUTE_CONFIGURATION || patch.ApplyTo == network.EnvoyFilter_HTTP_ROUTE {
+				// provide an error message indicating a mismatch between the operation type and the filter type
+				message := msg.NewEnvoyFilterUsesAddOperationIncorrectly(r)
 
-	// if the envoyFilter does not have a priority and it uses the INSERT_BEFORE or INSERT_AFTER
-	// then it can have issues if its using an operation that is relative to another filter.
-	// the default priority for an envoyFilter is 0
-	if ef.Priority == 0 {
-		for index, patch := range ef.ConfigPatches {
-			if patch.Patch.Operation == network.EnvoyFilter_Patch_INSERT_BEFORE || patch.Patch.Operation == network.EnvoyFilter_Patch_INSERT_AFTER {
-				// indicate that this envoy filter does not have a priority and has a relative patch
-				// operation set which can cause issues. Indicate a warning to the use to use a
-				// priority to ensure its placed after something or use the INSERT_FIRST option to
-				//  ensure its always done first
-				message := msg.NewEnvoyFilterUsesRelativeOperation(r)
+				if line, ok := util.ErrorLine(r, fmt.Sprintf(util.EnvoyFilterConfigPath, index)); ok {
+					message.Line = line
+				}
+
+				c.Report(collections.IstioNetworkingV1Alpha3Envoyfilters.Name(), message)
+			}
+		} else if patch.Patch.Operation == network.EnvoyFilter_Patch_REMOVE {
+			// the REMOVE operation is ignored when applyTo is set to ROUTE_CONFIGURATION, or HTTP_ROUTE.
+			if patch.ApplyTo == network.EnvoyFilter_ROUTE_CONFIGURATION || patch.ApplyTo == network.EnvoyFilter_HTTP_ROUTE {
+				// provide an error message indicating a mismatch between the operation type and the filter type
+				message := msg.NewEnvoyFilterUsesRemoveOperationIncorrectly(r)
 
 				if line, ok := util.ErrorLine(r, fmt.Sprintf(util.EnvoyFilterConfigPath, index)); ok {
 					message.Line = line
 				}
 
 				c.Report(collections.IstioNetworkingV1Alpha3Envoyfilters.Name(), message)
+			} else {
+				// A relative operation (REMOVE) was used so check if priority is set and if not set provide a warning
+				relativeOperationMsg(r, c, index, ef.Priority)
+			}
+		} else if patch.Patch.Operation == network.EnvoyFilter_Patch_REPLACE {
+			// the REPLACE operation is only valid for HTTP_FILTER and NETWORK_FILTER.
+			if patch.ApplyTo != network.EnvoyFilter_NETWORK_FILTER && patch.ApplyTo != network.EnvoyFilter_HTTP_FILTER {
+				// provide an error message indicating an invalid filter type
+				message := msg.NewEnvoyFilterUsesReplaceOperationIncorrectly(r)
+
+				if line, ok := util.ErrorLine(r, fmt.Sprintf(util.EnvoyFilterConfigPath, index)); ok {
+					message.Line = line
+				}
+
+				c.Report(collections.IstioNetworkingV1Alpha3Envoyfilters.Name(), message)
+			} else {
+				count := 0
+				for _, name := range patchFilterNames {
+					if instanceName.String() == name {
+						count++
+						break
+					}
+				}
+				if count > 0 {
```

Comment:
This count check doesn't seem quite right to me. Can you explain what exactly you're trying to do?

Reply:
was just trying to indicate if there is a patch being applied to the same envoy filter being used then should provide the warning indicating if a relative operation is being used.  If there is no previous envoy filter then don't need point out a relative is being used because then it would not matter since it will just get applied.

Reply:
Hmm, the doc for REPLACE says: `If the named filter is not found, this operation has no effect.` which is the same thing that happens if the matched filter for INSERT_BEFORE/AFTER is not found, but in those cases you do give a warning. Seems you should also warn here as well. Also, if the named filter is there, but after this one (instead of before) that seems like an error. Not sure how much you really should be checking here, but I think it should at least be consistent.

Reply:
concerned if we are always giving warnings it will get ignored so was trying to narrow what  cases to indicate when its important to note things are relative.  Reason I didn't mark INSERT_BEFORE/AFTER was because the doc didn't mention what happened if the named filter was not found.  The doc mentions that if no filter is provided then with INSERT_BEFORE/AFTER then it will go either first or last.  From the issue it sounded like the doc was unclear if the filter is mentioned but not yet applied.  So it seemed we needed a warning since the doc was unclear as to behavior.  Since the doc explicitly says for REPLACE that the operation is ignored then was thinking  the behavior was known so no warning was needed.

Reply:
if you think narrowing the cases is too confusing can also remove the checks so its always indicating a warning

Reply:
okay made another check for just the proxyVersion so that if that was the issue it would stand out more while leaving the other warning messages.  Theory being that maybe more warnings are better and if we provide a different warning for the proxyVersion it will stand out more